### PR TITLE
UofT-DSI | Deep Learning - Labs

### PR DESCRIPTION
**Pull request by sehroz**
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)

I am making adjustments to the code to improve functionality, fix errors, or implement new features.

## What did you learn from the changes you have made?

I learned that the changes addressed the issue at hand, improved the functionality, or clarified the errors encountered in the process.

## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?

Yes, I considered alternative methods (e.g., changing how files are handled, adjusting the logic) but chose the most efficient or appropriate solution for the task.

## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?

Yes, I faced challenges like debugging errors, handling data properly, or ensuring the correct file paths. I overcame them by troubleshooting, checking error messages, and adjusting the code accordingly.

## How were these changes tested?

The changes were tested by running the code with various inputs and verifying the outputs against expected results.

## A reference to a related issue in your repository (if applicable)

N/A

## Checklist
- [x] I can confirm that my changes are working as intended

### CODE REVIEW

File: 01_materials/labs/lab_1.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_1.ipynb:1]: When testing model trained with different `lr`, please declare variables under different names or run the training procedure in the `compile-fit` order. In your current version, model is overridden by the last compiled instance with `lr=0.001`.

File: 01_materials/labs/lab_2.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_2.ipynb:1]: Could you comment on why your best accuracy in the hyper-parameter tuning exercise was lower than the one you trained before?

File: 01_materials/labs/lab_3.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_3.ipynb:1]: * Could you explain the difference between the improved model in the last exercise and the baseline model you implemented before?
* It seems that there are some unresolved error in the section where you're finding the most similar films to "Star Wars" and "Star Trek".  This was likely caused by an issue identified during the live learning session with the first code block. Could you please follow the code below or follow the recording to fix the issue and re-run the notebook?

Original:
```
if not ML_100K_FOLDER.exists():
    if not ML_100K_FILENAME.exists():
        urlretrieve(ML_100K_URL, ML_100K_FILENAME.name)
    with ZipFile(ML_100K_FILENAME.name) as zip:
        zip.extractall()
```
Change to:
```
if not ML_100K_FOLDER.exists():
    if not ML_100K_FILENAME.exists():
        urlretrieve(ML_100K_URL, ML_100K_FILENAME.name)
    with ZipFile(ML_100K_FILENAME.name) as zipf:
        zipf.extractall()
```

File: 01_materials/labs/lab_5.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_5.ipynb:1]: Please double check the version you submitted. There were a few `#TODO`s left blank in the `Build a better model` session.

File: 01_materials/labs/lab_4.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_4.ipynb:1]: Please double check the version you submitted. The `Pooling and strides with convolutions` section was not completed.

File: 01_materials/labs/lab_6.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_6.ipynb:1]: Please double check the version you submitted. The `Preprocessing text for the (supervised) CBOW model` section was not completed.

**Review by tianyi21 — CHANGES_REQUESTED** [review]
Hi Sehroz!

Thank you for submitting your DL Workbooks. Please find my comments attached to each file. 

Please double check the file you submitted since a few sections were not complete in your Lab3-6. Please revise and re-submit your Lab3-6 accordingly. If you have any questions, please don't hesitate to let us know.

Thanks,
Tianyi [LS]

**Review by tianyi21 — APPROVED** [review]
Hi Sehroz!

Thank you for revising your DL Workbooks. Your workbooks are now complete.

All your DL assignments have been completed. Thank you for your participation, and I wish you all the best for your future!

Thanks,
Tianyi [LS]

File: 01_materials/labs/lab_1.ipynb
Comment (sehroz) [reply 01_materials/labs/lab_1.ipynb:1]: I've updated it prevent model override by creating separate instances for each learning rate test.

File: 01_materials/labs/lab_2.ipynb
Comment (sehroz) [reply 01_materials/labs/lab_2.ipynb:1]: Fixed the use of training data instead of test data in hyperparameter tuning. Model was incorrectly training on X_test rather than X_train, leading to poor accuracy.

File: 01_materials/labs/lab_3.ipynb
Comment (sehroz) [reply 01_materials/labs/lab_3.ipynb:1]: Fixed extending recommendation model with additional layers and hyperparameters. Modified final exercise model beyond baseline by adding dense layers, dropout, larger embeddings and MSE loss.

File: 01_materials/labs/lab_4.ipynb
Comment (sehroz) [reply 01_materials/labs/lab_4.ipynb:1]: Implemented max and average pooling visualization.

File: 01_materials/labs/lab_5.ipynb
Comment (sehroz) [reply 01_materials/labs/lab_5.ipynb:1]: Implemented object detection model with dual classification and localization heads using CNN architecture.

File: 01_materials/labs/lab_6.ipynb
Comment (sehroz) [reply 01_materials/labs/lab_6.ipynb:1]: Completed the preprocessing text for the CBOW model section as requested and made the necessary revisions.
